bittide/src/Bittide/ObliviousFifo.hs

```diff
@@ -0,0 +1,224 @@
+                    ) =>
+                    BitVector n ->
+                    BitVector (1 + Max n (BitSize a))
+                ) $ pack n
+    Data x ->
+      case leMaxRight @n @(BitSize a) @0 of
+        Dict ->
+          ( zeroExtend ::
+              ( (1 + Max n (BitSize a) - BitSize a) + BitSize a
+              ~ 1 + Max n (BitSize a)
+              ) =>
+              BitVector (BitSize a) ->
+              BitVector (1 + Max n (BitSize a))
+          ) $ pack x
+
+  unpack bv
+    | testBit bv (natToNum @(Max n (BitSize a))) =
+        let tbv = case leMaxLeft @n @(BitSize a) @0 of
+                    Dict ->
+                      ( truncateB ::
+                          ( n + (1 + Max n (BitSize a) - n)
+                          ~ 1 + Max n (BitSize a)
+                          ) =>
+                          BitVector (1 + Max n (BitSize a)) ->
+                          BitVector n
+                      ) bv
+        in case unpack tbv of
+          0 -> Empty
+          n -> Lost n
+    | otherwise = Data $ unpack $
+        case leMaxRight @n @(BitSize a) @0 of
+          Dict ->
+            ( truncateB ::
+                ( BitSize a + (1 +  Max n (BitSize a) - BitSize a)
+                ~ 1 + Max n (BitSize a)
+                ) =>
+                BitVector (1 + Max n (BitSize a)) ->
+                BitVector (BitSize a)
+            ) bv
+
+-- | This FIFO combines 'Clash.Cores.Xilinx.DcFifo.dcFifo' with
+-- 'Bittide.Counter.domainDiffCounter' to create an oblivious FIFO
+-- that supports a much larger _virtual_ data counter than the number
+-- of elements that it can actually hold. The FIFO is able to hold
+-- @2^n-1@ virtual elements and @2^m-1@ real data elements, where @n@
+-- can be much larger than @m@. When its real size @m@ is exceeded,
+-- then the FIFO does not loose the frames, but only forgets about
+-- their data. This means, that the frames are still counted,
+-- "traverse" the FIFO and are output at the other end in the same
+-- order as they are written to the component. The only difference is
+-- that frames, which are written to the FIFO, when its real capacity
+-- is already utilized, are output with a @lost@ status instead.
+--
+-- Due to the maximum depth of 17 of
+-- 'Clash.Cores.Xilinx.DcFifo.dcFifo', the current implementation
+-- requires the real FIFO to be not continuously full for more than
+-- 2^16-1 (= 65535) cycles. Note that this would require a read domain
+-- whose clock is 65535 times faster than the one of the write
+-- domain. Hence, this should only rarely be a limitation in practice.
+
+{-# NOINLINE xilinxObliviousFifo #-}
+xilinxObliviousFifo ::
+  forall n m readDom writeDom a.
+  ( HasCallStack
+  , KnownDomain readDom
+  , KnownDomain writeDom
+  , NFDataX a
+  , KnownNat n  -- the FIFO is able to hold 2^n-1 virtual elements
+  , KnownNat m  -- the FIFO is able to hold 2^m-1 real data elements
+  , 4 <= m, m <= 17
+  , 16 <= n, n <= 65
+  ) =>
+  SNat m ->
```

Comment:
What happens to this `m`? I don't see it mentioned anywhere else.

Reply:
Good point. It is meant to be size of the Xilinx FIFO data elements, which currently is fixed to `16`. There was a restriction in `Clash.Cores.Xilinx.DcFifo` which required this fix. You can make it dynamic, I just didn't want to update the `dcFifo` for the "proof of concept" yet.

Reply:
Yes that's fair

Reply:
The `m` is from a legacy version and not present any more.